src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1095,16 +1095,14 @@ private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request
                 throw CreateRetryException();
             }
 
```

Comment:
GitHub won't let me comment on it directly, but the above logic catches an ObjectDisposedException.  That's bothered me since it was added.  Can it be removed now?

Reply:
I looked at this but unfortunately I think there's still an issue here.

There race that's alluded to here is like this:
(1) Pool retrieves connection
(2) Connection receives GOAWAY, sets _lastStreamId to indicate we're shutting down and connection cannot be used for new requests, aborts existing streams that won't be processed so they will be retried, etc.
(3) All requests complete, we call CheckForShutdown, shutdown happens and we dispose stuff like _headerSerializationLock and _concurrentStreams
(4) Pool now tries to use connection, hits this logic

I think we can fix all this, but it's not correct as it stands. I will look at it further.

BTW, this also means we can't currently fully remove the active waiters logic from the credit manager. If I can fix this, I'll remove that too.